### ISSUE TITLE
work around XDG noncompliance causing "Unable to make or open a FIFO...: permission denied" error

### ISSUE THREAD

**Issue by jab** [fish-shell/fish-shell#2222]
@ridiculousfish, saw your [SO comment](http://stackoverflow.com/questions/27615899/fish-error-upon-startup-to-access-a-file-that-does-exist#comment43702109_27615899) after [hitting this myself](https://github.com/Homebrew/linuxbrew/issues/482#issuecomment-122331158) and couldn't find a corresponding ticket, so figured I'd open this in case there's still interest.

**Comment by jab** [comment]
In case it's helpful, I hit this after sshing into an Ubuntu 14.04.2 system as root, then running `su -l ubuntu` (where the ubuntu user's shell is /bin/bash), and then running `fish` (or `fish -l`), where `fish` is the linuxbrew-built version 2.2.0 in `~ubuntu/.linuxbrew/bin/fish`.

**Comment by faho** [comment]
Soo... I'll try to localize the information here. ~~Interrupt~~ reply to me when I'm wrong.

Your issue is that your fish on ubuntu 14.04 (LTS) can't access $XDG_RUNTIME_DIR, which is set to "/run/user/0".

XDG_RUNTIME_DIR is usually `/run/user/(id -u)` (where `id -u` is the UID), which means this is root's (which seems logical given it's also owned by root).

The issues appear before fish fully starts, so you can't set the variable in config.fish (presumably that's in $XDG_CONFIG_HOME, though IIRC the standard dictates a ~/.config fallback).

`/run/user/(id -u)` exists and is owned by you?

This is weird. I'm not quite sure what su implementation ubuntu is using and how they inject that variable into the environment (it's usually done via a pam module IIUC), but the [man page for util-linux su](http://www.man7.org/linux/man-pages/man1/su.1.html) says:

>       -, -l, --login
              Start the shell as a login shell with an environment similar
              to a real login:

>                 o      **clears all the environment variables** except TERM

>                o      **initializes the environment variables HOME, SHELL,
                        USER, LOGNAME, and PATH**

>                 o      changes to the target user's home directory

>                 o      sets argv[0] of the shell to '-' in order to make
                        the shell a login shell

So this means the shell should be getting the variable anew from somewhere else, and not just inheriting it from bash. The annoying thing is, fish can deal with it being unset, it just can't deal with a wrong value - which it arguable should.

In conclusion, there are two issues here:

1. fish should fallback to another path (preferably whatever XDG says) also if the variable is wrong

2. Your/Ubuntu's configuration is borked by setting a wrong variable

If I were you, I'd try to workaround it by setting XDG_RUNTIME_DIR manually.

**Comment by jab** [comment]
This does look like a bug in Ubuntu's su; thanks for the explanation and workaround. I briefly looked for an upstream bug report but couldn't immediately find one.

**Comment by zanchey** [comment]
Did you end up asking on Launchpad at all?

**Comment by jab** [comment]
I didn't. I also no longer have access to the server I was hitting this on and haven't hit it since.

**Comment by zanchey** [comment]
Should be fixed with #2470!

**Comment by GuillaumeLeclerc** [comment]
@zanchey 

Indeed it's fixed thanks !
